File: backend/src/main/java/org/smarttrainer/backend/modules/seance/service/SeanceService.java
package org.smarttrainer.backend.modules.seance.service;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.smarttrainer.backend.domain.coach.CoachProfile;
import org.smarttrainer.backend.domain.programme.Programme;
import org.smarttrainer.backend.domain.seance.Seance;
import org.smarttrainer.backend.modules.coach.repository.CoachProfileRepository;
import org.smarttrainer.backend.modules.programme.repository.ProgrammeRepository;
import org.smarttrainer.backend.modules.seance.dto.SeanceRequest;
import org.smarttrainer.backend.modules.seance.dto.SeanceResponse;
import org.smarttrainer.backend.modules.seance.mapper.SeanceMapper;
import org.smarttrainer.backend.modules.seance.repository.SeanceRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SeanceService {

    private final SeanceRepository seanceRepository;
    private final CoachProfileRepository coachRepository;
    private final ProgrammeRepository programmeRepository;
    private final SeanceMapper seanceMapper;

    @Transactional
    public SeanceResponse create(SeanceRequest request) {
        CoachProfile coach = coachRepository.findById(request.getCoachId())
                .orElseThrow(() -> new RuntimeException("Coach not found"));

        Programme programme = programmeRepository.findById(request.getProgrammeId())
                .orElseThrow(() -> new RuntimeException("Programme not found"));

        Seance seance = seanceMapper.toEntity(request);
        seance.setCoach(coach);
        seance.setProgramme(programme);

        return seanceMapper.toResponse(seanceRepository.save(seance));
    }

    public SeanceResponse getById(Long id) {
        Seance seance = seanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Seance not found"));
        return seanceMapper.toResponse(seance);
    }

    public List<SeanceResponse> getAll() {
        return seanceRepository.findAll()
                .stream()
                .map(seanceMapper::toResponse)
                .collect(Collectors.toList());
    }

    public List<SeanceResponse> getByCoachId(Long coachId) {
        return seanceRepository.findByCoachId(coachId)
                .stream()
                .map(seanceMapper::toResponse)
                .collect(Collectors.toList());
    }

    public List<SeanceResponse> getByProgrammeId(Long programmeId) {
        return seanceRepository.findByProgrammeId(programmeId)
                .stream()
                .map(seanceMapper::toResponse)
                .collect(Collectors.toList());
    }

    public List<SeanceResponse> getByStatus(String status) {
        return seanceRepository.findByStatus(status)
                .stream()
                .map(seanceMapper::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public SeanceResponse update(Long id, SeanceRequest request) {
        Seance seance = seanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Seance not found"));

        seance.setStatus(request.getStatus());
        seance.setDuration(request.getDuration());
        seance.setNotes(request.getNotes());

        return seanceMapper.toResponse(seanceRepository.save(seance));
    }

    @Transactional
    public void delete(Long id) {
        if (!seanceRepository.existsById(id)) {
            throw new RuntimeException("Seance not found");
        }
        seanceRepository.deleteById(id);
    }
}

File: frontend/src/app/core/services/seance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

// ── Response DTO ───────────────────────────────────────────────────────────────

export interface SeanceResponse {
  id: number;
  createdAt?: string;
  status: string;
  duration: number;       // en minutes
  notes?: string;
  coachId: number;
  coachName?: string;
  programmeId: number;
  programmeTitle?: string;
}

// ── Request DTO ────────────────────────────────────────────────────────────────

export interface SeanceRequest {
  status: string;
  duration: number;
  notes?: string;
  coachId: number;
  programmeId: number;
}

// ── Service ────────────────────────────────────────────────────────────────────

@Injectable({
  providedIn: 'root'
})
export class SeanceService {
  private apiUrl = 'http://localhost:8082/api/v1/seances';

  constructor(private http: HttpClient) { }

  getAll(): Observable<SeanceResponse[]> {
    return this.http.get<SeanceResponse[]>(this.apiUrl);
  }

  getById(id: number): Observable<SeanceResponse> {
    return this.http.get<SeanceResponse>(`${this.apiUrl}/${id}`);
  }

  getByCoachId(coachId: number): Observable<SeanceResponse[]> {
    return this.http.get<SeanceResponse[]>(`${this.apiUrl}/coach/${coachId}`);
  }

  getByProgrammeId(programmeId: number): Observable<SeanceResponse[]> {
    return this.http.get<SeanceResponse[]>(`${this.apiUrl}/programme/${programmeId}`);
  }

  getByStatus(status: string): Observable<SeanceResponse[]> {
    return this.http.get<SeanceResponse[]>(`${this.apiUrl}/status/${status}`);
  }

  create(request: SeanceRequest): Observable<SeanceResponse> {
    return this.http.post<SeanceResponse>(this.apiUrl, request);
  }

  update(id: number, request: SeanceRequest): Observable<SeanceResponse> {
    return this.http.put<SeanceResponse>(`${this.apiUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
